Extended Convergence system to steady-state detection

joshuahansel opened the pull request:
- Allowed Convergence objects to be used for steady-state detection
- Added `Convergence::checkIterationType(IterationType)`

joshuahansel committed:
Deleted deprecated steady check parameters

Refs #30858

joshuahansel committed:
Refactored solution diff norm in TransientBase

Refs #30858

joshuahansel committed:
Improved parameter descriptions in TransientMultiApp

Refs #30858

joshuahansel committed:
Added DefaultSteadyConvergence

Closes #30858

joshuahansel committed:
Added iteration type checks for Convergence objects

Closes #28927

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3032376/), step Clang format on 6a0c3da wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31010/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31010/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format abd55dd9c84792b711f8443be52289c38b18ff17`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3033544/), step Generate coverage on 609c8b4 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f9ceee</th>
      <th colspan=4>#31010 609c8b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.53%</td>
      <td>85.52%</td>
      <td>-0.01%</td>
      <td>81.01%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>115312</td>
      <td>115414</td>
      <td>+102</td>
      <td>145</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19514</td>
      <td>19540</td>
      <td>+26</td>
      <td>34</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31010/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31010/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>f9ceee</th>
      <th colspan=4>#31010 609c8b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.00%</td>
      <td>85.99%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8179</td>
      <td>8178</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1331</td>
      <td>1332</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31010/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31010/coverage/fluid_properties)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>f9ceee</th>
      <th colspan=4>#31010 609c8b</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.28%</td>
      <td>85.26%</td>
      <td>-0.03%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>19226</td>
      <td>19220</td>
      <td>-6</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3318</td>
      <td>3324</td>
      <td>+6</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31010/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31010/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31010/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 81.01% is less than the suggested 90.0%
- `navier_stokes` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud reviewed (changes requested):
will finish tomorrow

joshuahansel committed:
Renamed "SteadyConvergence" to "SteadyStateConvergence"

Refs #30858

joshuahansel committed:
Suppressed "enable" in Executioner

joshuahansel committed:
Fixed Convergence tests

Refs #30858

joshuahansel commented on framework/src/executioners/TransientBase.C:
```diff
@@ -760,39 +763,18 @@ TransientBase::setTimeStepper(TimeStepper & ts)
-    // This line is useful since _solution_change_norm will be printed
-    _solution_change_norm =
-        *std::max_element(aux_soln_change_norms.begin(), aux_soln_change_norms.end());
+    _console << "Steady-state Convergence object reported divergence" << std::endl;
```
Yeah, I'll make it an error.

joshuahansel committed:
Threw error if steady-state convergence diverges

Refs #30858

joshuahansel commented on framework/include/executioners/TransientBase.h:
```diff
@@ -304,4 +296,12 @@ class TransientBase : public Executioner
   void constrainDTFromMultiApp(Real & dt_cur,
                                std::ostringstream & diag,
                                const ExecFlagType & execute_on) const;
+
+  /// Whether to use the auxiliary system solution to determine steady-states
+  const bool _check_aux;
+
+  /// Whether to divide the solution difference norm by dt. If taking 'small' time steps this member
+  /// should probably be true. If taking very 'large' timesteps in an attempt to reach a
+  /// steady-state, this member should probably be be false.
+  const bool _normalize_solution_diff_norm_by_dt;
```
This leftover stuff is mostly because of `TransientMultiApp`, which can do SS checks on the multiapps during sub-cycling. Unfortunately, it offers its own tolerance value, which implies I probably won't be able to get away with just using the multiapp's SS Convergence object to do the check, but I will make the change and see what happens in testing.

joshuahansel replied:
Looks like just one test in framework failing from this change. It was fixed by setting the steady-state tolerance in the sub input file to match the one in the MultiApp block in the parent.

Do you see any issues with requiring all steady-state detection criteria to be specified in the sub app? I just don't know why we currently have a tolerance in TransientMultiApp. I'm thinking that parameter should go away, so I suggest deprecation.

joshuahansel replied:
I do find that setting `detect_steady_state = true` in `TransientMultiApp` is not the same as keeping it false and then setting `steady_state_detection = true` in the sub, so that's probably relevant here.

joshuahansel committed:
Removed SS detection member data from TransientBase

Refs #30858

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/3054252/), step Clang format on 037ee59 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/31010/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/31010/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2bf80888ed8064ebb4f92fba1db5c743ac2be9dd`

joshuahansel commented on modules/phase_field/test/tests/initial_conditions/MultiSmoothSuperellipsoidIC_2D.i:
```diff
@@ -51,7 +51,6 @@
   start_time = 0.0
   num_steps = 1
   dt = 100.0
-  enable = false
```
Suppressed it now

joshuahansel commented on test/tests/convergence/default_multiapp_fixed_point_convergence/parent.i:
```diff
@@ -77,10 +77,10 @@ T_ambient = 300.0
+  # nl_max_its = 10
+  # nl_abs_tol = 1e-8
+  # nl_rel_tol = 1e-8
+  # l_tol = 1e-3
```
Deleted

joshuahansel committed:
Removed "steady_state_tol" from TransientMultiApp

Refs #30858

GiudGiud commented on framework/include/convergence/Convergence.h:
```diff
@@ -37,10 +37,25 @@ class Convergence : public MooseObject,
+    LINEAR = 1, // solve_type = LINEAR, not linear solves within nonlinear solve
+    MULTISYSTEM_FIXED_POINT = 2,
+    MULTIAPP_FIXED_POINT = 3,
+    STEADY_STATE = 4
```
we should consider either using a MooseEnum (and storing it somewhere) to which an app could add, OR adding a CUSTOM option for apps to leverage

GiudGiud replied:
this could be an idea for later too btw

this isnt top priority rn

GiudGiud commented on framework/src/convergence/DefaultSteadyStateConvergence.C:
```diff
@@ -0,0 +1,77 @@
+    norm = *std::max_element(aux_var_diff_norms.begin(), aux_var_diff_norms.end());
+  }
+  else
+    norm = _transient_executioner->computeSolutionChangeNorm(_check_aux, _normalize_norm_by_dt);
```
why have check_aux here if you are calling different code for check_aux = true

GiudGiud commented on framework/src/executioners/Steady.C:
```diff
@@ -36,6 +36,8 @@ Steady::init()
   _problem.execute(EXEC_PRE_MULTIAPP_SETUP);
   _problem.initialSetup();
+  _fixed_point_solve->initialSetup();
+  _feproblem_solve.initialSetup();
```
so the executioner is in charge of calling initial setup on the problem and the solve objects.

are we doing the same for solve() / postSolve() / ?

joshuahansel replied:
The only other option I see is making `Executioner::initialSetup()` and then calling that in `FEProblemBase::initialSetup()`, since the executioner owns the solves.

joshuahansel replied:
Appears consistent to me.

GiudGiud commented on framework/src/executioners/TransientBase.C:
```diff
@@ -36,10 +37,39 @@
+  params.addParam<bool>("check_aux",
+                        false,
+                        "Whether to check the auxiliary system for convergence to steady-state. If "
+                        "false, then the nonlinear system is used.");
```
```suggestion
                        "false, then the solution vector from the solver systems is used.");
```

GiudGiud commented on framework/src/multiapps/TransientMultiApp.C:
```diff
@@ -49,15 +49,9 @@ TransientMultiApp::validParams()
-                        "considered to be at steady state");
+                        "If true, then if/while sub-cycling ('sub_cycling = true'), a steady-state "
+                        "check will be performed for each child app, allowing them to skip to the "
+                        "end of the parent time step if steady conditions are detected.");
```
so if you set this, it makes the child apps create SteadyState cv checks.
but the other parameter is just cumbersome to make happen?

GiudGiud replied:
> I don't understand what you mean, but I think this parameter can be eliminated

If we don't use it, works for me

joshuahansel commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2524,6 +2556,8 @@ class FEProblemBase : public SubProblem, public Restartable
   std::optional<std::vector<ConvergenceName>> _linear_convergence_names;
   /// MultiApp fixed point convergence name
   std::optional<ConvergenceName> _multiapp_fixed_point_convergence_name;
+  /// Steady convergence name
```
I'm going to use language "steady-state detection convergence object"

GiudGiud replied:
> It feels more problem-y, but there's no TransientProblem.

Right that is the problem. Should there be some sort of helper class to isolate these transient-only things?

joshuahansel committed:
Addressed review

- Improved "check_aux" doc
- Used "false" instead of "_check_aux" when always false
- Improved steady-state detection method doxygen

Refs #30858

joshuahansel committed:
Added Convergence iteration type registry

Refs #30858

GiudGiud commented on framework/include/utils/ConvergenceIterationTypeEnum.h:
fits better under convergence/ than utils?

GiudGiud replied:
so does the other file imo

GiudGiud commented on framework/src/convergence/DefaultSteadyStateConvergence.C:
```diff
@@ -0,0 +1,78 @@
+  InputParameters params = DefaultConvergenceBase::validParams();
+  params += TransientBase::defaultSteadyStateConvergenceParams();
+
+  params.addClassDescription("Default steady-state convergence criteria.");
```
```suggestion
  params.addClassDescription("Default convergence criteria for detecting steady-state conditions in a transient simulation.");
```

GiudGiud commented on framework/src/convergence/DefaultSteadyStateConvergence.C:
```diff
@@ -0,0 +1,78 @@
+  TIME_SECTION(_perfid_check_convergence);
+
+  Real norm = 0.0;
+  if (_check_aux)
+  {
+    std::vector<Number> aux_var_diff_norms;
+    _aux_system.variableWiseRelativeSolutionDifferenceNorm(aux_var_diff_norms);
+    for (auto & aux_var_diff_norm : aux_var_diff_norms)
+      aux_var_diff_norm /= (_normalize_norm_by_dt ? _fe_problem.dt() : Real(1));
+
+    norm = *std::max_element(aux_var_diff_norms.begin(), aux_var_diff_norms.end());
+  }
+  else
+    norm = _transient_executioner->computeSolutionChangeNorm(false, _normalize_norm_by_dt);
```
```suggestion
  norm = _transient_executioner->computeSolutionChangeNorm(_check_aux, _normalize_norm_by_dt);
```
would this work?

joshuahansel replied:
`_transient_executioner->computeSolutionChangeNorm(true, _normalize_norm_by_dt)` expands to:
```
_aux.solution().l2_norm_diff(_aux.solutionOld()) / _aux.solution().l2_norm() / (_normalize_norm_by_dt ? _dt : Real(1));
```
so it doesn't do it variable-wise.

GiudGiud replied:
it does solution-vector wise

so inconsistent between aux and solver systems then

joshuahansel replied:
Yes, unfortunately. Not sure why it happened, but that's why having a Convergence system is nice 😀

GiudGiud replied:
agreed. Let's add a note to the documentation though so as to let the people know of our quirks

GiudGiud commented on framework/src/executioners/TransientBase.C:
```diff
@@ -208,18 +197,27 @@ TransientBase::TransientBase(const InputParameters & parameters)
 
   _time_stepper->init();
 
+  auto & conv = _problem.getConvergence(_problem.getSteadyStateConvergenceName());
```
so we always have one? even when we are not looking to detect steady state? 
I think this is in line with the prior code that always computed `_solution_change_norm` but I dont think we need to

GiudGiud reviewed (commented):
looks about ready to me

joshuahansel committed:
Moved ConvergenceIterationTypeEnum

Refs #30858

joshuahansel committed:
Improved DefaultSteadyStateConvergence description

Refs #30858

joshuahansel committed:
Addressed review

- Added Convergence iteration registry documentation
- Also added missing includes
- Added note about not needing steady-state Convergence always
- Added note about different aux/solution behavior
- Moved ConvergenceIterationTypes

Refs #30858
